README.rst
.. -*- mode: rst; coding: utf-8 -*-

.. image:: pypiserver_logo.png
   :width: 300 px
   :align: center

==============================================================================
pypiserver - minimal PyPI server for use with pip/easy_install
==============================================================================
|pypi-ver| |travis-status| |dependencies| |python-ver| |proj-license|

:Version:     1.2.5
:Date:        2018-11-11
:Source:      https://github.com/pypiserver/pypiserver
:PyPI:        https://pypi.org/project/pypiserver/
:Travis:      https://travis-ci.org/pypiserver/pypiserver
:Maintainers: Kostis Anagnostopoulos <ankostis@gmail.com>,
              Matthew Planchard <mplanchard@gmail.com>
:License:     zlib/libpng + MIT

*pypiserver* is a minimal PyPI_ compatible server for *pip* or *easy_install*.
It is based on bottle_ and serves packages from regular directories.
Wheels, bdists, eggs and accompanying PGP-signatures can be uploaded
either with *pip*, *setuptools*, `twine`_, `pypi-uploader`_, or simply copied
with *scp*.


.. contents:: Table of Contents
  :backlinks: top


Quickstart: Installation and Usage
==================================
*pypiserver* ``> 1.2.x`` works with Python ``2.7`` and ``3.4+`` or *pypy*.
Older Python versions may still work, but they are not tested.
For legacy Python versions, use ``pypiserver-1.1.x`` series.

.. Tip::
   The commands below work on a unix-like operating system with a posix shell.
   The ``'~'`` character expands to user's home directory.

   If you're using Windows, you'll have to use their "Windows counterparts".
   The same is true for the rest of this documentation.


1. Install *pypiserver* with this command::

    pip install pypiserver                ## Or: pypiserver[passlib,watchdog]
    mkdir ~/packages                      ## Copy packages into this directory.

   See also `Alternative Installation methods`_.

2. Copy some packages into your ``~/packages`` folder and then
   get your *pypiserver* up and running::

    pypi-server -p 8080 ~/packages &      ## Will listen to all IPs.

3. From the client computer, type this::

    ## Download and Install hosted packages.
    pip install  --extra-index-url http://localhost:8080/pypi/web/simple/ ...

    # or
    pip install --extra-index-url http://localhost:8080

    ## Search hosted packages
    pip search --index http://localhost:8080 ...

    # Note that pip search does not currently work with the /pypi/web/simpleendpoint

   See also `Client-side configurations`_ for avoiding tedious typing.

4. Enter ``pypi-server -h`` in the cmd-line to print a detailed usage message::

    pypi-server [OPTIONS] [PACKAGES_DIRECTORY...]
      start PyPI compatible package server serving packages from
      PACKAGES_DIRECTORY. If PACKAGES_DIRECTORY is not given on the
      command line, it uses the default ~/packages.  pypiserver scans this
      directory recursively for packages. It skips packages and
      directories starting with a dot. Multiple package directories can be
      specified.

    pypi-server understands the following options:

      -p, --port PORT
        listen on port PORT (default: 8080)

      -i, --interface INTERFACE
        listen on interface INTERFACE (default: 0.0.0.0, any interface)

      -a, --authenticate (UPDATE|download|list), ...
        comma-separated list of (case-insensitive) actions to authenticate
        Use '.' or '' for empty. Requires to have set the password (-P option).
        For example to password-protect package downloads (in addition to uploads)
        while leaving listings public, give:
          -P foo/htpasswd.txt  -a update,download
        To drop all authentications, use:
          -P .  -a .
        Note that when uploads are not protected, the `register` command
        is not necessary, but `~/.pypirc` still need username and password fields,
        even if bogus.
        By default, only 'update' is password-protected.

      -P, --passwords PASSWORD_FILE
        use apache htpasswd file PASSWORD_FILE to set usernames & passwords when
        authenticating certain actions (see -a option).
        If you want to allow un-authorized access, set this option and -a
        explicitly to empty (either '.' or'').

      --disable-fallback
        disable redirect to real PyPI index for packages not found in the
        local index

      --fallback-url FALLBACK_URL
        for packages not found in the local index, this URL will be used to
        redirect to (default: https://pypi.org/pypi/web/simple)

      --server METHOD
        use METHOD to run the server. Valid values include paste,
        cherrypy, twisted, gunicorn, gevent, wsgiref, auto. The
        default is to use "auto" which chooses one of paste, cherrypy,
        twisted or wsgiref.

      -r, --root PACKAGES_DIRECTORY
        [deprecated] serve packages from PACKAGES_DIRECTORY

      -o, --overwrite
        allow overwriting existing package files

      --hash-algo ALGO
        any `hashlib` available algo used as fragments on package links.
        Set one of (0, no, off, false) to disabled it. (default: md5)

      --welcome HTML_FILE
        uses the ASCII contents of HTML_FILE as welcome message response.

      -v
        enable INFO logging;  repeat for more verbosity.

      --log-conf <FILE>
        read logging configuration from FILE.
        By default, configuration is read from `log.conf` if found in server's dir.

      --log-file <FILE>
        write logging info into this FILE.

      --log-frmt <FILE>
        the logging format-string.  (see `logging.LogRecord` class from standard python library)
        [Default: %(asctime)s|%(name)s|%(levelname)s|%(thread)d|%(message)s]

      --log-req-frmt FORMAT
        a format-string selecting Http-Request properties to log; set to  '%s' to see them all.
        [Default: %(bottle.request)s]

      --log-res-frmt FORMAT
        a format-string selecting Http-Response properties to log; set to  '%s' to see them all.
        [Default: %(status)s]

      --log-err-frmt FORMAT
        a format-string selecting Http-Error properties to log; set to  '%s' to see them all.
        [Default: %(body)s: %(exception)s \n%(traceback)s]

    pypi-server -h
    pypi-server --help
      show this help message

    pypi-server --version
      show pypi-server's version

    pypi-server -U [OPTIONS] [PACKAGES_DIRECTORY...]
      update packages in PACKAGES_DIRECTORY. This command searches
      pypi.org for updates and shows a pip command line which
      updates the package.

    The following additional options can be specified with -U:

      -x
        execute the pip commands instead of only showing them

      -d DOWNLOAD_DIRECTORY
        download package updates to this directory. The default is to use
        the directory which contains the latest version of the package to
        be updated.

      -u
        allow updating to unstable version (alpha, beta, rc, dev versions)

    Visit https://github.com/pypiserver/pypiserver for more information.



Client-side Configurations
==========================
Always specifying the the pypi url on the command line is a bit
cumbersome. Since *pypiserver* redirects ``pip/easy_install`` to the
``pypi.org`` index if it doesn't have a requested package, it's a
good idea to configure them to always use your local pypi index.

Configuring *pip*
-----------------
For ``pip`` command this can be done by setting the environment variable
``PIP_EXTRA_INDEX_URL`` in your ``.bashr/.profile/.zshrc``::

  export PIP_EXTRA_INDEX_URL=http://localhost:8080/pypi/web/simple/

or by adding the following lines to ``~/.pip/pip.conf``::

  [global]
  extra-index-url = http://localhost:8080/pypi/web/simple/

.. Note::
   If you have installed *pypiserver* on a remote url without *https*
   you wil receive an "untrusted" warning from *pip*, urging you to append
   the ``--trusted-host`` option.  You can also include this option permanently
   in your configuration-files or environment variables.


Configuring *easy_install*
--------------------------
For ``easy_install`` command you may set the following configuration in
``~/.pydistutils.cfg``::

  [easy_install]
  index_url = http://localhost:8080/pypi/web/simple/


Uploading Packages Remotely
===========================
Instead of copying packages directly to the server's folder (i.e. with ``scp``),
you may use python tools for the task, e.g. ``python setup.py upload``.
In that case, *pypiserver* is responsible for authenticating the upload-requests.


.. Note::
  We strongly advise to password-protected your uploads!

  It is possible to disable authentication for uploads (e.g. in intranets).
  To avoid lazy security decisions, read help for ``-P`` and ``-a`` options.


*Apache*-like authentication (``htpasswd``)
-------------------------------------------
#. First make sure you have the *passlib* module installed (note that
   ``passlib>=1.6`` is required), which is needed for parsing the Apache
   *htpasswd* file specified by the ``-P``, ``--passwords`` option
   (see next steps)::

     pip install passlib

#. Create the Apache *htpasswd* file with at least one user/password pair
   with this command (you'll be prompted for a password)::

     htpasswd -sc htpasswd.txt <some_username>

   .. Tip:: Read this SO question for running `htpasswd` cmd
      under *Windows*:

         http://serverfault.com/questions/152950/how-to-create-and-edit-htaccess-and-htpasswd-locally-on-my-computer-and-then-u

      or if you have bogus passwords that you don't care because they are for
      an internal service (which is still "bad", from a security prespective...)
      you may use this public service:

         http://www.htaccesstools.com/htpasswd-generator/

   .. Tip:: When accessing pypiserver via the api, alternate authentication
      methods are available via the ``auther`` config flag. Any callable
      returning a boolean can be passed through to the pypiserver config in
      order to provide custom authentication. For example, to configure
      pypiserver to authenticate using the `python-pam`_::

        import pam
        pypiserver.default_config(auther=pam.authenticate)

      Please see `Using Ad-hoc authentication providers`_ for more information.

#. You  need to restart the server with the ``-P`` option only once
   (but user/password pairs can later be added or updated on the fly)::

     ./pypi-server -p 8080 -P htpasswd.txt ~/packages &

Upload with *setuptools*
------------------------
#. On client-side, edit or create a ``~/.pypirc`` file with a similar content::

     [distutils]
     index-servers =
       pypi
       local

     [pypi]
     username:<your_pypi_username>
     password:<your_pypi_passwd>

     [local]
     repository: http://localhost:8080
     username: <some_username>
     password: <some_passwd>

#. Then from within the directory of the python-project you wish to upload,
   issue this command::

     python setup.py sdist upload -r local


Upload with `twine`_
--------------------
To avoid storing you passwords on disk, in clear text, you may either:

- use the ``register`` *setuptools*'s command with the ``-r`` option,
  like that::

     python setup.py sdist register -r local upload -r local

- use `twine`_ library, which
  breaks the procedure in two steps.  In addition, it supports signing
  your files with PGP-Signatures and uploading the generated `.asc` files
  to *pypiserver*::

     twine upload -r local --sign -identity user_name ./foo-1.zip


Upload with `pypi-uploader`_
----------------------------
You can also upload packages using `pypi-uploader`_, which
obviates the need to download packages locally prior to uploading them to
pypiserver. You can install it with ``pip install pypi-uploader``, and
assuming you have a ``pypi_local`` source set up in your ``~/.pypirc``,
use it like this::

    pypiupload packages mock==1.0.1 requests==2.2.1 -i pypi_local
    pypiupload requirements requirements.txt -i pypi_local


Using the Docker Image
======================

Starting with version 1.2.5, official Docker images will be built for each
push to master, each dev, alpha, or beta release, and each final release.
The most recent full release will always be available under the tag ``latest``,
and the current master branch will always be available under the tag
``unstable``.

You can always check to see what tags are currently available at our
`Docker Repo`_.

To run the most recent release of ``pypiserver`` with Docker, simply::

    docker run pypiserver/pypiserver:latest

This starts ``pypiserver`` serving packages from the ``/data/packages``
directory inside the container, listening on the container port 8080.

The container takes all the same arguments as the normal ``pypi-server``
executable, with the exception of the internal container port (``-p``),
which will always be 8080.

Of course, just running a container isn't that interesting. To map
port 80 on the host to port 8080 on the container::

    docker run -p 80:8080 pypiserver/pypiserver:latest

You can now access your ``pypiserver`` at ``localhost:80`` in a web browser.

To serve packages from a directory on the host, e.g. ``~/packages``::

    docker run -p 80:8080 -v ~/packages:/data/packages pypiserver/pypiserver:latest

To authenticate against a local ``.htpasswd`` file::

    docker run -p 80:8080 -v ~/.htpasswd:/data/.htpasswd pypiserver/pypiserver:latest -P .htpasswd packages

You can also specify ``pypiserver`` to run as a Docker service using a
composefile. An example composefile is `provided <docker-compose.yml>`_.


.. _`docker repo`: https://hub.docker.com/r/pypiserver/pypiserver/tags/


Alternative Installation Methods
================================
When trying the methods below, first use the following command to check whether
previous versions of *pypiserver* already exist, and (optionally) uninstall them::

  ## VERSION-CHECK: Fails if not installed.
  pypi-server --version

  ## UNINSTALL: Invoke again untill it fails.
  pip uninstall pypiserver


Installing the very latest version
----------------------------------
In case the latest version in *pypi* is a pre-release, you have to use
*pip*'s `--pre` option.  And to update an existing installation combine it
with `--ignore-installed`::

  pip install pypiserver --pre -I

You can even install the latest *pypiserver* directly from *github* with the
following command, assuming you have *git* installed on your ``PATH``::

  pip install git+git://github.com/pypiserver/pypiserver.git


Installing it as standalone script
----------------------------------
The git repository contains a ``pypi-server-standalone.py`` script,
which is a single python file that can be executed without any other
dependencies.

Run the following commands to download the script with ``wget``::

  wget https://raw.github.com/pypiserver/pypiserver/standalone/pypi-server-standalone.py
  chmod +x pypi-server-standalone.py

or with ``curl``::

  curl -O https://raw.github.com/pypiserver/pypiserver/standalone/pypi-server-standalone.py
  chmod +x pypi-server-standalone.py

You can then start-up the server with::

  ./pypi-server-standalone.py

Feel free to rename the script and move it into your ``$PATH``.


Running on *heroku/dotcloud*
----------------------------
https://github.com/dexterous/pypiserver-on-the-cloud contains
instructions on how to run *pypiserver* on one of the supported cloud
service providers.



Recipes
=======

Managing the package directory
------------------------------
The ``pypi-server`` command has the ``-U`` option that searches for updates of
available packages. It scans the package directory for available
packages and searches on pypi.org for updates. Without further
options ``pypi-server -U`` will just print a list of commands which must
be run in order to get the latest version of each package. Output
looks like::

    $ ./pypi-server -U
    checking 106 packages for newer version

    .........u.e...........e..u.............
    .....e..............................e...
    ..........................

    no releases found on pypi for PyXML, Pymacs, mercurial, setuptools

    # update raven from 1.4.3 to 1.4.4
    pip -q install --no-deps  --extra-index-url https://pypi.org/pypi/web/simple-d /home/ralf/packages/mirror raven==1.4.4

    # update greenlet from 0.3.3 to 0.3.4
    pip -q install --no-deps  --extra-index-url https://pypi.org/pypi/web/simple-d /home/ralf/packages/mirror greenlet==0.3.4

It first prints for each package a single character after checking the
available versions on pypi. A dot(`.`) means the package is up-to-date, ``'u'``
means the package can be updated and ``'e'`` means the list of releases on
pypi is empty. After that it shows a *pip* command line which can be used
to update a one package. Either copy and paste that or run
``pypi-server -Ux`` in order to really execute those commands. You need
to have *pip* installed for that to work however.

Specifying an additional ``-u`` option will also allow alpha, beta and
release candidates to be downloaded. Without this option these
releases won't be considered.


Serving thousands of packages
-----------------------------

By default, *pypiserver* scans the entire packages directory each time an
incoming HTTP request occurs. This isn't a problem for a small number of
packages, but causes noticeable slow-downs when serving thousands of packages.

If you run into this problem, significant speedups can be gained by enabling
pypiserver's directory caching functionality. The only requirement is to
install the ``watchdog`` package, or it can be installed during ``pypiserver``
installation, by specifying the ``cache`` extras option::

    pip install pypiserver[cache]

If you are using a static webserver such as *Apache* or *nginx* as
a reverse-proxy for pypiserver, additional speedup can be gained by
directly serving the packages directory:

For instance, in *nginx* you may adding the following config to serve
packages-directly directly (take care not to expose "sensitive" files)::

    location /packages/ {
      root /path/to/packages/parentdir;
    }

If you have packages that are very large, you may find it helpful to
disable hashing of files (set ``--hash-algo=off``, or ``hash_algo=None`` when
using wsgi).


Managing Automated Startup
--------------------------
There are a variety of options for handling the automated starting of
pypiserver upon system startup. Two of the most common are *systemd* and
*supervisor*.


Running as a *systemd* service
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*systemd* is installed by default on most modern Linux systems and as such,
it is an excellent option for managing the pypiserver process. An example
config file for ``systemd`` can be seen below::

    [Unit]
    Description=A minimal PyPI server for use with pip/easy_install.
    After=network.target

    [Service]
    Type=simple
    # systemd requires absolute path here too.
    PIDFile=/var/run/pypiserver.pid
    User=www-data
    Group=www-data

    ExecStart=/usr/local/bin/pypi-server -p 8080 -a update,download --log-file /var/log/pypiserver.log -P /etc/nginx/.htpasswd /var/www/pypi
    ExecStop=/bin/kill -TERM $MAINPID
    ExecReload=/bin/kill -HUP $MAINPID
    Restart=always

    WorkingDirectory=/var/www/pypi

    TimeoutStartSec=3
    RestartSec=5

    [Install]
    WantedBy=multi-user.target

Adjusting the paths and adding this file as ``pypiserver.service`` into your
``systemd/system`` directory will allow management of the pypiserver process with
``systemctl``, e.g. ``systemctl start pypiserver``.

More useful information about *systemd* can be found at
https://www.digitalocean.com/community/tutorials/how-to-use-systemctl-to-manage-systemd-services-and-units


Launching through *supervisor*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

`supervisor <http://supervisord.org/>`_ has the benefit of being a pure python
package and as such, it provides excellent cross-platform support for process
management. An example configuration file for ``supervisor`` is given below::

    [program:pypi]

CHANGES.rst
Changelog
=========

1.2.5 (2018-11-11)
------------------

- FIX: bcrypt is now installed into the Docker image, which allows
  passlib to work like it should (thanks @Diftraku, #224)

- MAINT: integration tests with ``twine`` have been updated to use the
  command-line interface rather than the internal API, which should
  make them more resilient over time (#226)

1.2.4 (2018-08-06)
------------------

- FIX: the command to download new versions of available packages now
  works with ``pip`` >= 10.0 (thanks @elboerto, #215)

1.2.3 (2018-08-04)
------------------

- MAINT: Remove broken downloads badge (thanks @hugovk, #209)

- ENH: Improved Dockerfile and ``docker-compose`` example, docs for using
  the docker image, automatic docker builds

1.2.2 (2018-06-12)
------------------

- FIX: update fallback URL to https://pypi.org/pypi/web/simplesince pypi.python.org
  has shut down

- FIX: updated tests to use ``Popen`` rather than ``pip.main()`` given its
  removal in pip version 10.0

- DOC: scrubbed docs of links to pypi.python.org

- DEPRECATION: Drop support for Python 3.3 (thanks @hugovk, #198)


1.2.1 (2017-11-29)
------------------

- FIX propagation of certain ``pypiserver`` settings via a ``paste.ini`` config
  file (thanks @luismsgomes, #156)

- FIX update default fallback URL to be https for compliance with PyPI
  (thanks @uSpike, #182)

- FIX resolved a regression preventing spinning up multiple pypiservers
  via a paste config (thanks @bertjwregeer, #173)

- FIX cmdline parsing of stray comparison consuming many flags (e.g. ``--help``),
  and docs about ``auther``
  - (thanks to @sakurai-youhei, #162).

- Travis CI testing for Python 3.6 and pypy3 (#183)

- Several documentation improvements (thanks @tescalada, #166, #161, #172 and
  @axnsan12, #190)

1.2.0 (2016-06-25)
------------------
"Brexit": Normalize and stop legacy support.

- Less rigorous support for ``python-2 < 2.7`` and ``python-3 < 3.3``.
- Package normalizations and :pep:`503` updates:
  - Package names are normalized: convert all characters to lower-case
    and replace any of ``[-_.]`` with a dash(``'-'``).
  - The simple index only lists normalized package names.
  - Any request for a non-normalized package name is redirected to
    the normalized name.
  - URLs are redirected unless they end in ``'/'`` (expect packages themselves).
  - (thanks to @dpkp, #38, #139, #140)

- Added ``pip search`` support.
  - (thanks to @blade2005, #80, #114)

- FIX startup regressions for other WSGI-servers, introduced by previous ``v1.1.10``.
  - (thanks to @virtuald, @Oneplus, @michaelkuty, @harcher81, @8u1a,
    #117, #122, #124/#127/#128)

- FIX over-writing of packages even when without ``--overwrite`` flag.
  - (thanks to @blade2005, #113)

- Fixes for *paste*, *gunicorn* and other *WSGI* servers.
  - (thanks to @corywright, @virtuald, @montefra, #112, #118, #119)

- Updates and fixes needed due to changes in dependent libraries.
  - (thanks @dpkp, #120/#121, #129, #141/#142)

- Add cache for speeding up GPG signatures.
  - sthanks to @virtuald, #116)

- Other minor fixes and improvements.
  - (thanks to @bibby, @Oneplus, @8u1a, #129, #131)

- TravisCI-test against *python-3.5*.
  - (#107, #108, #110)

- docs:

  - Provide samples for *Automated Startup* (``systemd`` & ``hypervisor``).
    (thanks to @ssbarnea, #137, #146)

  - Add usage instructions for related project ``pypi-uploader``.
    (thanks to @ssbarnea & @bibby, #147)

  - doc: Provide sample-code to authenticate using ``/etc/passwds`` file
    via *pam* modules in Unix.
    - (thanks to @blade2005, #149, #151-#153)

  - Improved API usage instructions.
  - Detailed changes recorded in `Github's milestone 1.2.0
    <https://github.com/pypiserver/pypiserver/milestones/M1.2.0>`_.


1.1.10 (2016-01-19)
-------------------
Serve 1000s of packages, PGP-Sigs, skip versions starting with 'v'.

+ #101: Speed-up server by (optionally) using the `watchdog` package
  to cache results, serve packages directly from proxying-server (*Apache* ,
  *nginx*), and pre-compile regexes (thanks @virtuald).
- #106: Support uploading PGP-signatures (thanks @mplanchard).
- Package-versions parsing modifications:

  - #104: Stopped parsing invalid package-versions prefixed with `v`; they are
    invalid according to :pep-reference:`0440` (thanks @virtuald &
    @stevejefferiesIDBS).
  - Support versions with epochs separated by `!` like `package-1!1.1.0`.
  - #102: FIX regression on uploading packages with `+` char in their version
    caused by recent bottle-upgrade.
- #103: Minor doc fixes (thanks @MichaelSchneeberger).


1.1.9 (2015-12-21)
------------------
"Ssss-elections" bug-fix & maintenance release.

- Upgrade bottle 1.11.6-->1.13-dev.

  - Fixes `MAX_PARAM` limiting dependencies(#82)

- Rework main startup and standalone:

  - New standalone generation based on ZIPed wheel archive.
  - Replace all sys.module mechanics with relative imports.
  - Fix gevent monkeypatching (#49).
  - Simplify definition of config-options on startup.
  - TODO: Move startup-options validations out of `main()` and
    into `pypiserver.core`
    package, to validate also start-up from API-clients.

- #53: Like PyPI, HREF-links now contain package's md5-hashes in their fragment.
  Add `--hash_algo` cmd-line option to turn-off or specify other *hashlib*
  message-digest algorithms (e.g. `sha256` is a safer choice, set it to `off`
  to avoid any performance penalty if hosting a lot of packages).

- #97: Add `--auther` non cmd-line startup-option to allow for alternative
  authentication methods (non HtPasswdFile-based one) to be defined by
  API-clients (thanks @Tythos).

- #91: Attempt to fix register http failures (thanks to @Tythos and @petri).

  - Test actual clients (ie `pip`, `Twine`, `setuptools`).
  - Test spurious `setuptools` failures.
  - NOT FIXED!  Still getting spurious failures.

- Various other fixes:

  - #96: Fix program's requirement (i.e. add passlib as extra-requirement).

pypiserver/__main__.py
#! /usr/bin/env python
"""
.. NOTE:: To the developer:
    This module is moved to the root of the standalone zip-archive,
    to be used as its entry-point. Therefore DO NOT import relative.
"""
from __future__ import print_function

import getopt
import logging
import os
import re
import sys
import textwrap

import functools as ft


log = logging.getLogger('pypiserver.main')


def init_logging(level=None, frmt=None, filename=None):
    logging.basicConfig(level=level, format=frmt)
    rlog = logging.getLogger()
    rlog.setLevel(level)
    if filename:
        rlog.addHandler(logging.FileHandler(filename))


def usage():
    return textwrap.dedent("""\
  pypi-server [OPTIONS] [PACKAGES_DIRECTORY...]
    start PyPI compatible package server serving packages from
    PACKAGES_DIRECTORY. If PACKAGES_DIRECTORY is not given on the
    command line, it uses the default ~/packages.  pypiserver scans this
    directory recursively for packages. It skips packages and
    directories starting with a dot. Multiple package directories can be
    specified.

  pypi-server understands the following options:

    -p, --port PORT
      listen on port PORT (default: 8080)

    -i, --interface INTERFACE
      listen on interface INTERFACE (default: 0.0.0.0, any interface)

    -a, --authenticate (UPDATE|download|list), ...
      comma-separated list of (case-insensitive) actions to authenticate
      Use '.' or '' for empty. Requires to have set the password (-P option).
      For example to password-protect package downloads (in addition to uploads)
      while leaving listings public, give:
        -P foo/htpasswd.txt  -a update,download
      To drop all authentications, use:
        -P .  -a .
      Note that when uploads are not protected, the `register` command
      is not necessary, but `~/.pypirc` still need username and password fields,
      even if bogus.
      By default, only 'update' is password-protected.

    -P, --passwords PASSWORD_FILE
      use apache htpasswd file PASSWORD_FILE to set usernames & passwords when
      authenticating certain actions (see -a option).
      If you want to allow un-authorized access, set this option and -a
      explicitly to empty (either '.' or'').

    --disable-fallback
      disable redirect to real PyPI index for packages not found in the
      local index

    --fallback-url FALLBACK_URL
      for packages not found in the local index, this URL will be used to
      redirect to (default: https://pypi.python.org/pypi/web/simple)

    --server METHOD
      use METHOD to run the server. Valid values include paste,
      cherrypy, twisted, gunicorn, gevent, wsgiref, auto. The
      default is to use "auto" which chooses one of paste, cherrypy,
      twisted or wsgiref.

    -r, --root PACKAGES_DIRECTORY
      [deprecated] serve packages from PACKAGES_DIRECTORY

    -o, --overwrite
      allow overwriting existing package files

    --hash-algo ALGO
      any `hashlib` available algo used as fragments on package links.
      Set one of (0, no, off, false) to disabled it. (default: md5)

    --welcome HTML_FILE
      uses the ASCII contents of HTML_FILE as welcome message response.

    -v
      enable verbose logging;  repeat for more verbosity.

    --log-file <FILE>
      write logging info into this FILE.

    --log-frmt <FILE>
      the logging format-string.  (see `logging.LogRecord` class from standard python library)
      [Default: %(asctime)s|%(name)s|%(levelname)s|%(thread)d|%(message)s]

    --log-req-frmt FORMAT
      a format-string selecting Http-Request properties to log; set to  '%s' to see them all.
      [Default: %(bottle.request)s]

    --log-res-frmt FORMAT
      a format-string selecting Http-Response properties to log; set to  '%s' to see them all.
      [Default: %(status)s]

    --log-err-frmt FORMAT
      a format-string selecting Http-Error properties to log; set to  '%s' to see them all.
      [Default: %(body)s: %(exception)s \n%(traceback)s]

    --cache-control AGE
      Add "Cache-Control: max-age=AGE, public" header to package downloads.
      Pip 6+ needs this for caching.


  pypi-server -h
  pypi-server --help
    show this help message

  pypi-server --version
    show pypi-server's version

  pypi-server -U [OPTIONS] [PACKAGES_DIRECTORY...]
    update packages in PACKAGES_DIRECTORY. This command searches
    pypi.org for updates and shows a pip command line which
    updates the package.

  The following additional options can be specified with -U:

    -x
      execute the pip commands instead of only showing them

    -d DOWNLOAD_DIRECTORY
      download package updates to this directory. The default is to use
      the directory which contains the latest version of the package to
      be updated.

    -u
      allow updating to unstable version (alpha, beta, rc, dev versions)

  Visit https://pypi.org/project/pypiserver/ for more information.
  """)


def main(argv=None):
    import pypiserver

    if argv is None:
        argv = sys.argv

    command = "serve"

    c = pypiserver.Configuration(**pypiserver.default_config())

    update_dry_run = True
    update_directory = None
    update_stable_only = True

    try:
        opts, roots = getopt.getopt(argv[1:], "i:p:a:r:d:P:Uuvxoh", [
            "interface=",
            "passwords=",
            "authenticate=",
            "port=",
            "root=",
            "server=",
            "fallback-url=",
            "disable-fallback",
            "overwrite",
            "hash-algo=",
            "log-file=",
            "log-frmt=",
            "log-req-frmt=",
            "log-res-frmt=",
            "log-err-frmt=",
            "welcome=",
            "cache-control=",
            "version",
            "help"
        ])
    except getopt.GetoptError:
        err = sys.exc_info()[1]
        sys.exit("usage error: %s" % (err,))

    for k, v in opts:
        if k in ("-p", "--port"):
            try:
                c.port = int(v)
            except Exception:
                err = sys.exc_info()[1]
                sys.exit("Invalid port(%r) due to: %s" % (v, err))
        elif k in ("-a", "--authenticate"):
            c.authenticated = [a.lower()
                               for a in re.split("[, ]+", v.strip(" ,"))
                               if a]
            if c.authenticated == ['.']:
                c.authenticated = []
            else:
                actions = ("list", "download", "update")
                for a in c.authenticated:
                    if a not in actions:
                        errmsg = "Action '%s' for option `%s` not one of %s!"
                        sys.exit(errmsg % (a, k, actions))
        elif k in ("-i", "--interface"):
            c.host = v
        elif k in ("-r", "--root"):
            roots.append(v)
        elif k == "--disable-fallback":
            c.redirect_to_fallback = False
        elif k == "--fallback-url":
            c.fallback_url = v
        elif k == "--server":
            c.server = v
        elif k == "--welcome":
            c.welcome_file = v
        elif k == "--version":
            print("pypiserver %s\n" % pypiserver.__version__)
            return
        elif k == "-U":
            command = "update"
        elif k == "-x":
            update_dry_run = False
        elif k == "-u":
            update_stable_only = False
        elif k == "-d":
            update_directory = v
        elif k in ("-P", "--passwords"):
            c.password_file = v
        elif k in ("-o", "--overwrite"):
            c.overwrite = True
        elif k in "--hash-algo":
            c.hash_algo = None if not pypiserver.str2bool(v, c.hash_algo) else v
        elif k == "--log-file":
            c.log_file = v
        elif k == "--log-frmt":
            c.log_frmt = v
        elif k == "--log-req-frmt":
            c.log_req_frmt = v
        elif k == "--log-res-frmt":
            c.log_res_frmt = v
        elif k == "--log-err-frmt":
            c.log_err_frmt = v
        elif k == "--cache-control":
            c.cache_control = v
        elif k == "-v":
            c.verbosity += 1
        elif k in ("-h", "--help"):
            print(usage())
            sys.exit(0)

    if (not c.authenticated and c.password_file != '.' or
            c.authenticated and c.password_file == '.'):
        auth_err = "When auth-ops-list is empty (-a=.), password-file (-P=%r) must also be empty ('.')!"
        sys.exit(auth_err % c.password_file)

    if len(roots) == 0:
        roots.append(os.path.expanduser("~/packages"))

    roots=[os.path.abspath(x) for x in roots]
    c.root = roots

    verbose_levels=[
        logging.WARNING, logging.INFO, logging.DEBUG, logging.NOTSET]
    log_level=list(zip(verbose_levels, range(c.verbosity)))[-1][0]
    init_logging(level=log_level, filename=c.log_file, frmt=c.log_frmt)

    if command == "update":
        from pypiserver.manage import update_all_packages
        update_all_packages(roots, update_directory,
                dry_run=update_dry_run, stable_only=update_stable_only)
        return

    # Fixes #49:
    #    The gevent server adapter needs to patch some
    #    modules BEFORE importing bottle!
    if c.server and c.server.startswith('gevent'):
        import gevent.monkey  # @UnresolvedImport
        gevent.monkey.patch_all()

    from pypiserver import bottle
    if c.server not in bottle.server_names:
        sys.exit("unknown server %r. choose one of %s" % (
            c.server, ", ".join(bottle.server_names.keys())))

    bottle.debug(c.verbosity > 1)
    bottle._stderr = ft.partial(pypiserver._logwrite,
            logging.getLogger(bottle.__name__), logging.INFO)
    app = pypiserver.app(**vars(c))
    bottle.run(app=app, host=c.host, port=c.port, server=c.server)


if __name__ == "__main__":
    main()
